[NRBF] Don't use Unsafe.As when decoding DateTime(s)

adamsitnik opened the pull request:
fixes #102826

adamsitnik committed:
add tests

adamsitnik committed:
fix

stephentoub commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs:
```diff
@@ -70,36 +74,67 @@ internal static object ReadPrimitiveValue(this BinaryReader reader, PrimitiveTyp
+#if NET
+        [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
+        extern static DateTime CallPrivateSerializationConstructor(SerializationInfo si, StreamingContext ct);
+#endif
```
Using a private constructor like this from a separate package is considered safe / supported?

jkotas replied:
I agree. `System.Formats.Nrbf` should not depend on any of the built-in legacy infrastructure for binary serialization.

stephentoub replied:
> AFAIK we have no plans to remove these ctors.

We've already made some of them throw PlatformNotSupportedException, e.g.
https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs#L175-L176
and entirely removed them from others, e.g.
https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs#L12

jkotas replied:
Should we have a public API that allows us to create the specific data time value?

adamsitnik replied:
> Should we have a public API that allows us to create the specific data time value?

Yes, but it would not solve the problem as this package needs to support older monikers, including `netstandard2.0`

jkotas replied:
I think both Unsafe.As and reflection are ok for existing targets. The existing targets are set in stone and we can make assumptions about them.

Unsafe.As or reflection are less than ideal for future. They limit the changes we can do in future.

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SystemClassWithMembersAndTypesRecord.cs:
```diff
@@ -77,10 +77,10 @@ internal SerializationRecord TryToMapToUserFriendly()
         else if (MemberValues.Count == 2
             && HasMember("ticks") && HasMember("dateData")
-            && MemberValues[0] is long value && MemberValues[1] is ulong
+            && MemberValues[0] is long && MemberValues[1] is ulong dateData
```
No. "ticks" is "dataData" with ticks mask applied:

https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L1548

https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L1299-L1301

GrabYourPitchforks commented:
Hi all! There seems to be some confusion here, so I'll try to clarify.

For libraries which could execute on the .NET Framework runtime, we are not allowed to call APIs or access members which are undocumented. Using `Unsafe.As<...>(ref datetime)` as a struct overlay violates this policy. We were previously alerted to a similar problem in the SqlClient package and resolved it. See https://github.com/dotnet/SqlClient/pull/1647#issuecomment-1159055343 for additional context. Much of what I'm going to say here is basically a copy of what I posted there.

First, this requirement only applies to libraries compiled against netfx directly or against netstandard (if they can run on netfx). If you're compiled against any other TFM, go wild. :)

Second, the serialization ctor is in fact "publicly" documented by virtue of the fact that the type implements the `ISerializable` interface. The contract of that interface is that the type is required to implement the necessary serialization ctor - often in a nonpublic manner - and callers can absolutely rely on it being present. When running on the netfx runtime, the code sample I provided is fully supported, fully reliable, and meets all compliance requirements we're bound by.

Third, if you do continue to use a struct overlay for platforms other than netfx, ensure that on these platforms it is legal and safe for `DateTime` to be backed by any arbitrary bit pattern. We have been bitten in the past where overlaying atop types like `decimal` has led to security incidents because arbitrary untrusted bit patterns can violate an instance's internal invariants.

jkotas commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SystemClassWithMembersAndTypesRecord.cs:
```diff
@@ -77,10 +77,10 @@ internal SerializationRecord TryToMapToUserFriendly()
         else if (MemberValues.Count == 2
             && HasMember("ticks") && HasMember("dateData")
-            && MemberValues[0] is long value && MemberValues[1] is ulong
+            && MemberValues[0] is long && MemberValues[1] is ulong dateData
```
Is it safe to depend on the order of the fields in the payload? In other words, is the exact order of fields part of the BF contract for given type?

GrabYourPitchforks replied:
`SerializationInfo` doesn't officially document the order, but in practice it enumerates elements in the same order in which they're added, and some types are sensitive to this ordering. It's akin to how if a dictionary / hashtable changes the order of enumeration or if a sort routine changes the relative order of "equal" elements, things break.

jkotas replied:
I meant it in the connection with your other comment: Is the order considered a documented .NET Framework detail that it is ok to depend on; or is the order undocumented .NET Framework detail and we should not depend on it?

(My hunch would be that it should be the later.)

adamsitnik replied:
It was my mistake to rely on the order of fields, I am going to push a fix in a minute.

jkotas commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs:
```diff
@@ -70,36 +74,67 @@ internal static object ReadPrimitiveValue(this BinaryReader reader, PrimitiveTyp
+            {
+#pragma warning disable SYSLIB0050 // Type or member is obsolete
+                SerializationInfo si = new(typeof(DateTime), new FormatterConverter());
+                si.AddValue("ticks", 0L); // legacy value (serialized as long) - specify both just to be safe
```
How does specifying `tick` makes us safe?

I think it can only hide problems and produce invalid values instead of throwing an exception. I cannot think about a case where it actually helps.

adamsitnik replied:
I can not speak on the behalf of @GrabYourPitchforks (who authored the code), but my understanding is that initially the `SerializationInfo` for `DateTime` contained only `ticks` field. Later `dateData` was introduced, but the runtime kept emitting the old field in case it could be deserialized with older runtime.

I've double checked the code:

https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/datetime.cs#L388-L389

https://github.com/dotnet/runtime/blob/eb655cf84d8ebe8f52c83a4fab6ef577c4a536c6/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L837-L839

And you are right, in case this code were executed on a very old runtime, we would provide an invalid result. I've removed it and added comment.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into safeDateTimeDecoding

adamsitnik committed:
address code review feedback:

- don't rely on the order of fields, as it's an implementation details that may change
- don't specify "ticks" in the SerializationInfo

jkotas commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SystemClassWithMembersAndTypesRecord.cs:
```diff
@@ -75,29 +75,30 @@ ulong value when TypeNameMatches(typeof(UIntPtr)) => Create(new UIntPtr(value)),
                 return Create(new TimeSpan(ticks));
             }
         }
         else if (MemberValues.Count == 2
```
- This assumes that the BF format is set is stone and that new key/value pairs won't be added in future. Is it safe assumption to make?

- If somebody constructs malicious payload with extra TimeSpan, DateTime or Guid fields or with fields of unexpected type, this pattern match won't kick in, there won't be any exception thrown and we return the raw data. Is it the desired behavior for Nrbf reader? (As far as I can tell, the reader tends to throw on anything unexpected or invalid instead of accepting it silently.)

adamsitnik replied:
These are great questions.

`BinaryFormatter` can represent same primitive value using different record types based on the context.
In this case, when `DateTime` is the root object it is expressed as `SystemClassWithMembersAndTypesRecord` which is just a type name + key/value dictionary. In other cases, it can be represented as `MemberPrimitiveTypedRecord<T>` (or a raw 8 bytes).

My goal was to hide this from the end users and always map it to `PrimitiveTypeRecord<T>` so users don't need to become experts in this area.

> This assumes that the BF format is set is stone and that new key/value pairs won't be added in future.

If we ever extend the binary representation of given types, we may need to handle the versioning here.


> If somebody constructs malicious payload with extra TimeSpan, DateTime or Guid fields or with fields of unexpected type, this pattern match won't kick in, there won't be any exception thrown and we return the raw data. Is it the desired behavior for Nrbf reader?

It's allowed to create a type that is called `System.DateTime` and has different layout, in such cases we are going to return a `ClassRecord` and the users will need to handle it.


```csharp
SerializationRecord rootObject = NrbfDecoder.Decode(payload);
if (rootObject is PrimitiveTypeRecord<DateTime> primitiveRecord)
{
    // DateTime
}
else if (rootObject is ClassRecord classRecord)
{
    // something else
}
```

@jkotas this is just the way I see it, please let me know if something is not clear or some other changes are needed

jkotas replied:
> My goal was to hide this from the end users and always map it to PrimitiveTypeRecord<T>

You are not *always* mapping it to `PrimitiveTypeRecord<T>`.

You are only mapping it to `PrimitiveTypeRecord<T>` if the input has specific shape. You are not mapping it for all possible valid input shapes. For example, if the payload was produced by .NET Framework 1.x (I am sure there are a bunch of such payloads still alive in the wild), it will be missing `dateData` field and it is not going to mapped. However, the classic BF deserializer is going to handle it just fine. If somebody runs into this case, they will have to do double the work: They will need to handle both mapped and the non-mapped cases.

In general, I would expect the behavior to be either:
- 100% compatible with classic BF deserializer
- Exception to be thrown
